File: internal/repository/models/setting.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"

	"github.com/xxcheng123/cloudpan189-share/internal/pkgs/ptr"
	"gorm.io/gorm"
)

type Setting struct {
	ID          int64           `gorm:"primaryKey" json:"id"`
	Title       string          `gorm:"column:title;type:varchar(255);not null" json:"title"`
	EnableAuth  bool            `gorm:"column:enable_auth;type:tinyint(1);default:1" json:"enableAuth"`
	SaltKey     string          `gorm:"column:salt_key;type:varchar(255);not null" json:"-"`
	BaseURL     string          `gorm:"column:base_url;type:varchar(255);not null;default:''" json:"baseURL"`
	Initialized bool            `gorm:"column:initialized;type:tinyint(1);default:0" json:"initialized"`
	Addition    SettingAddition `gorm:"column:addition;type:json" json:"addition" swaggertype:"object"`
	CreatedAt   time.Time       `gorm:"column:created_at;autoCreateTime;type:datetime;default:CURRENT_TIMESTAMP" json:"createdAt"`
	UpdatedAt   time.Time       `gorm:"column:updated_at;autoUpdateTime;type:datetime;default:CURRENT_TIMESTAMP;on update:CURRENT_TIMESTAMP" json:"updatedAt"`
}

func (s *Setting) TableName() string { return "setting" }

func (s *Setting) AfterFind(tx *gorm.DB) (err error) {
	s.Addition.applyDefaults()
	return nil
}

type SettingAddition struct {
	Keep                      string `json:"-"`
	LocalProxy                bool   `json:"localProxy"`
	MultipleStream            bool   `json:"multipleStream"`
	MultipleStreamThreadCount int    `json:"multipleStreamThreadCount"`
	MultipleStreamChunkSize   int64  `json:"multipleStreamChunkSize"`
	TaskThreadCount           int    `json:"taskThreadCount"`

	ExternalAPIKey             string `json:"externalApiKey"`
	DefaultTokenId             int64  `json:"defaultTokenId"`
	ExternalAutoRefreshEnabled *bool  `json:"externalAutoRefreshEnabled"`
	ExternalRefreshIntervalMin int    `json:"externalRefreshIntervalMin"`
	ExternalAutoRefreshDays    int    `json:"externalAutoRefreshDays"`

	PersistentCheckEnabled bool   `json:"persistentCheckEnabled"`
	PersistentCheckDay     int    `json:"persistentCheckDay"`
	PersistentCheckTime    string `json:"persistentCheckTime"`

	AutoDeleteInvalidStorageEnabled  bool   `json:"autoDeleteInvalidStorageEnabled"`
	AutoDeleteInvalidStorageKeywords string `json:"autoDeleteInvalidStorageKeywords"`

	CasTargetEnabled            bool   `json:"casTargetEnabled"`
	CasTargetType               string `json:"casTargetType"`
	CasPersonTargetTokenId      int64  `json:"casPersonTargetTokenId"`
	CasPersonTargetFolderId     string `json:"casPersonTargetFolderId"`
	CasPersonAccessPath         string `json:"casPersonAccessPath"`
	CasFamilyTargetTokenId      int64  `json:"casFamilyTargetTokenId"`
	CasFamilyTargetFamilyId     string `json:"casFamilyTargetFamilyId"`
	CasFamilyTargetFolderId     string `json:"casFamilyTargetFolderId"`
	CasFamilyAccessPath         string `json:"casFamilyAccessPath"`
	CasRestoreRetentionHours    int    `json:"casRestoreRetentionHours"`
	RecycleBinAutoClearEnabled  bool   `json:"recycleBinAutoClearEnabled"`
	RecycleBinAutoClearTime     string `json:"recycleBinAutoClearTime"`
	LocalCASAutoScanEnabled     bool   `json:"localCasAutoScanEnabled"`
	LocalCASAutoScanIntervalMin int    `json:"localCasAutoScanIntervalMin"`
	CasAutoCollectEnabled       bool   `json:"casAutoCollectEnabled"`
	CasAutoCollectPreservePath  bool   `json:"casAutoCollectPreservePath"`
}

func (sa *SettingAddition) applyDefaults() {
	if sa.MultipleStreamThreadCount <= 0 {
		sa.MultipleStreamThreadCount = 4
	}
	if sa.MultipleStreamChunkSize <= 0 {
		sa.MultipleStreamChunkSize = 4 * 1024 * 1024
	}
	if sa.TaskThreadCount <= 0 {
		sa.TaskThreadCount = 1
	}
	if sa.ExternalAutoRefreshEnabled == nil {
		sa.ExternalAutoRefreshEnabled = ptr.Of(true)
	}
	if sa.ExternalRefreshIntervalMin <= 0 {
		sa.ExternalRefreshIntervalMin = 60
	}
	if sa.ExternalAutoRefreshDays <= 0 {
		sa.ExternalAutoRefreshDays = 60
	}
	if sa.PersistentCheckDay <= 0 || sa.PersistentCheckDay > 28 {
		sa.PersistentCheckDay = 1
	}
	if sa.PersistentCheckTime == "" {
		sa.PersistentCheckTime = "03:00"
	}
	if sa.AutoDeleteInvalidStorageKeywords == "" {
		sa.AutoDeleteInvalidStorageKeywords = "分享审核|分享审核不通过|没有找到分享信息|页面不存在|文件或文件夹不存在|分享平铺目录未找到"
	}
	if sa.CasTargetType == "" {
		sa.CasTargetType = "person"
	}
	if sa.RecycleBinAutoClearTime == "" {
		sa.RecycleBinAutoClearTime = "03:30"
	}
	if sa.LocalCASAutoScanIntervalMin <= 0 {
		sa.LocalCASAutoScanIntervalMin = 10
	}
	if !sa.CasAutoCollectPreservePath {
		sa.CasAutoCollectPreservePath = true
	}
}

func (sa SettingAddition) Value() (driver.Value, error) {
	if sa == (SettingAddition{}) {
		return nil, nil
	}
	return json.Marshal(sa)
}

func (sa *SettingAddition) Scan(value interface{}) error {
	if value == nil {
		*sa = SettingAddition{}
		return nil
	}
	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("cannot scan SettingAddition from non-string/[]byte value")
	}
	return json.Unmarshal(bytes, sa)
}

File: fe/src/api/setting.ts
import { api, type ApiResponse } from '@/utils/api'

// 系统初始化请求接口
export interface InitSystemRequest {
  baseURL: string // 系统基础URL
  enableAuth: boolean // 是否启用认证
  superUsername: string // 超级管理员用户名，长度3-20位
  superPassword: string // 超级管理员密码，长度6-20位
  title: string // 系统标题
}

// 修改系统标题请求接口
export interface ModifyTitleRequest {
  title: string // 新的系统标题
}

// 修改系统基础URL请求接口
export interface ModifyBaseURLRequest {
  baseURL: string // 新的系统基础URL
}

// ===== 系统设置接口 =====

// 获取系统信息
export const getSystemInfo = (): Promise<ApiResponse<Models.SystemInfo>> => {
  return api.get('/setting/info').then((res) => res.data)
}

// 初始化系统
export const initSystem = (data: InitSystemRequest): Promise<ApiResponse> => {
  return api.post('/setting/init_system', data).then((res) => res.data)
}

// 修改系统标题（管理员权限）
export const modifySystemTitle = (title: string): Promise<ApiResponse> => {
  const data: ModifyTitleRequest = { title }
  return api.post('/setting/modify_title', data).then((res) => res.data)
}

// 修改系统基础URL（管理员权限）
export const modifySystemBaseURL = (baseURL: string): Promise<ApiResponse> => {
  const data: ModifyBaseURLRequest = { baseURL }
  return api.post('/setting/modify_base_url', data).then((res) => res.data)
}

// 切换系统鉴权开关（管理员权限）
export interface ToggleEnableAuthRequest {
  enableAuth: boolean // 是否启用鉴权
}

export const toggleSystemEnableAuth = (enableAuth: boolean): Promise<ApiResponse> => {
  const data: ToggleEnableAuthRequest = { enableAuth }
  return api.post('/setting/toggle_enable_auth', data).then((res) => res.data)
}

// 查询系统附加设置（仅登录用户）
export const getSettingAddition = (): Promise<ApiResponse<Models.SettingAddition>> => {
  return api.get('/setting/addition').then((res) => res.data)
}

// 可选修改系统附加设置（管理员权限）
export interface ModifySettingAdditionRequest {
  localProxy?: boolean
  multipleStream?: boolean
  multipleStreamThreadCount?: number
  multipleStreamChunkSize?: number
  taskThreadCount?: number

  externalApiKey?: string
  defaultTokenId?: number
  externalAutoRefreshEnabled?: boolean
  externalRefreshIntervalMin?: number
  externalAutoRefreshDays?: number

  persistentCheckEnabled?: boolean
  persistentCheckDay?: number
  persistentCheckTime?: string

  autoDeleteInvalidStorageEnabled?: boolean
  autoDeleteInvalidStorageKeywords?: string

  casTargetEnabled?: boolean
  casTargetTokenId?: number
  casTargetType?: string
  casTargetFamilyId?: string
  casTargetFolderId?: string
  casAccessPath?: string
  casAutoCollectEnabled?: boolean
  casAutoCollectPreservePath?: boolean
}

export const modifySettingAddition = (data: ModifySettingAdditionRequest): Promise<ApiResponse> => {
  return api.post('/setting/modify_addition', data).then((res) => res.data)
}

export const clearCasTargetCache = (): Promise<ApiResponse<{ deleted: number }>> => {
  return api.post('/setting/clear_cas_target_cache').then((res) => res.data)
}

export const rebuildCasTargetCache = (): Promise<ApiResponse<{ dirCount: number; itemCount: number }>> => {
  return api.post('/setting/rebuild_cas_target_cache').then((res) => res.data)
}

export const runAutoDeleteInvalidStorageOnce = (): Promise<ApiResponse> => {
  return api.post('/setting/run_auto_delete_invalid_storage_once').then((res) => res.data)
}
